csr2transmart/csr_mapper.py

```
@@ -37,6 +42,96 @@ def row_value_to_value(concept_col, value_type: ValueType) -> Optional[Value]:
+        self.concept_key_to_modifier_key = concept_key_to_modifier_key
+        self.modifier_key_to_modifier = modifier_key_to_modifier
+        self.observations: List[Observation] = []
+
+    def map_observation_metadata(self, modifier_key, value: str) -> Optional[ObservationMetadata]:
+        if modifier_key is None or value is None:
+            return None
+        mod_metadata: Dict[Modifier, Value] = dict()
+        modifier = self.modifier_key_to_modifier.get(modifier_key)
+        if modifier is None:
+            return None
+        mod_metadata[modifier] = CategoricalValue(value)
+        return ObservationMetadata(mod_metadata)
+
+    def add_observations_for_entity(self, entity, entity_id: str, patient: Patient):
+        concept_keys = entity.fields.keys()
+        for concept_key in concept_keys:
+            concept = self.concept_key_to_concept.get(concept_key.upper())
+            if concept is not None:
+                value = row_value_to_value(getattr(entity, concept_key), concept.value_type)
+                if isinstance(entity, Individual):
+                    metadata = None
+                else:
+                    modifier_key = self.concept_key_to_modifier_key.get(concept_key.upper())
+                    metadata = self.map_observation_metadata(modifier_key, entity_id)
+                observation = Observation(patient, concept, None, self.default_trial_visit, None, None, value,
+                                          metadata)
+                self.observations.append(observation)
+
+    def map_individual_linked_entity_observations(self, entities: Sequence[Any], id_attribute: str):
+        for entity in entities:
+            entity_id = getattr(entity, id_attribute)
+            patient = self.individual_id_to_patient.get(entity.individual_id)
+            if patient is None:
+                logger.warning('No patient with identifier: {}. Skipping creating observation for {} with id: {}.'
```

I think indeed that in these cases it is better to raise an exception.